src/object/srv_ec_aggregate.c

```diff
@@ -0,0 +1,927 @@
+	unsigned int	as_extent_cnt;  /* number of replica extents         */
+	unsigned int	as_offset;      /* start offset in stripe            */
+	unsigned int	as_prefix_ext;  /* prefix range to delete            */
+};
+
+/* Aggregation state for an object. (may need to restructure if
+ * list of these is built from committed DTX table)
+ */
+struct ec_agg_entry {
+	d_list_t		 ae_link;	 /* List (of entries) link    */
+	daos_unit_oid_t		 ae_oid;	 /* OID of iteration entry    */
+	struct daos_oclass_attr	*ae_oca;	 /* Object class of object    */
+	struct obj_ec_codec	*ae_codec;	 /* Encode/decode for oclass  */
+	d_sg_list_t		*ae_sgl;	 /* Mem for entry processing  */
+	daos_handle_t		 ae_cont_hdl;    /* Container handle	      */
+	daos_handle_t		 ae_chdl;	 /* Vos container handle      */
+	daos_handle_t		 ae_thdl;	 /* Iterator handle           */
+	daos_epoch_range_t	 ae_epr;	 /* hi/lo extent threshold    */
+	daos_key_t		 ae_dkey;	 /* Current dkey              */
+	daos_key_t		 ae_akey;	 /* Current akey              */
+	daos_size_t		 ae_rsize;	 /* Record size of cur array  */
+	struct ec_agg_stripe	 ae_cur_stripe;  /* Struct for current stripe */
+	struct ec_agg_par_extent ae_par_extent;	 /* Parity extent             */
+	daos_handle_t		 ae_obj_hdl;	 /* Object handle for cur obj */
+};
+
+/* Struct used to drive offloaded stripe update.
+ */
+struct ec_agg_stripe_ud {
+	struct ec_agg_entry	*asu_agg_entry; /* Associated entry     */
+	uint8_t			*asu_bit_map;   /* Bitmap of cells      */
+	unsigned int		 asu_cell_cnt;  /* Count of cells       */
+	bool			 asu_recalc;    /* Should recalc parity */
+	ABT_eventual		 asu_eventual;  /* Eventual for offload */
+};
+
+/* Represents an replicated data extent.
+ */
+struct ec_agg_extent {
+	d_list_t	ae_link;  /* for extents list   */
+	daos_recx_t	ae_recx;  /* idx, nr for extent */
+	daos_epoch_t	ae_epoch; /* epoch for extent   */
+};
+
+/* Reset iterator state upon completion of iteration of a subtree.
+ */
+static inline void
+reset_agg_pos(vos_iter_type_t type, struct ec_agg_entry *agg_entry)
+{
+	switch (type) {
+	case VOS_ITER_DKEY:
+		memset(&agg_entry->ae_dkey, 0, sizeof(agg_entry->ae_dkey));
+		break;
+	case VOS_ITER_AKEY:
+		memset(&agg_entry->ae_akey, 0, sizeof(agg_entry->ae_akey));
+		break;
+	default:
+		break;
+	}
+}
+
+/* Compare function for keys.  Used to reset iterator position.
+ */
+static inline int
+agg_key_compare(daos_key_t key1, daos_key_t key2)
+{
+	if (key1.iov_len != key2.iov_len)
+		return 1;
+
+	return memcmp(key1.iov_buf, key2.iov_buf, key1.iov_len);
+}
+
+/* Handles dkeys returned by the per-object nested iteratior.
+ */
+static int
+agg_dkey(daos_handle_t ih, vos_iter_entry_t *entry,
+	 struct ec_agg_entry *agg_entry, unsigned int *acts)
+{
+	int rc = 0;
+
+	if (agg_key_compare(agg_entry->ae_dkey, entry->ie_key)) {
+		agg_entry->ae_dkey = entry->ie_key;
+		reset_agg_pos(VOS_ITER_AKEY, agg_entry);
+	}
+	agg_entry->ae_dkey	= entry->ie_key;
```

Comment:
This function and the duplicated dkey/akey actually doesn't provide any value in current implementation, since we can always use the dkey/akey in vos_iter_entry_t directly.

VOS aggregation saves last processed oid, dkey and akey to deal with the situation of: Assume iterator iterates into subtree and yield in subtree iteration, when it comes back to parent tree after the subtree iteration done, it has to do a re-probe since the parent tree might have been changed by others while it's yielding in subtree iteration. This re-probe will likely locate the last processed parent entry if it's still there (not being deleted by others), so we have to store the last processed oid/dkey/akey in aggregation param and compare it with current oid/dkey/akey when iterating each oid/dkey/akey entry, if it's same, we know that this entry has already been processed and inform vos_iterate() (by VOS_ITER_CB_SKIP flag) to avoid iterating into subtree again.

I think you have to deal with this situation for EC aggregation as well.